### PR TITLE
Fix/lp 1934614

### PR DESCRIPTION
**Pull request by ycliuhw**
*Application status of sidecar charms now get updated more accurately based on units status;*

driveby:

- rewrite the caasapplicationprovisioner test suite for easier adding additional tests;
- add missed tests for `Scale` and `Trust` logic in caasapplicationprovisioner worker;

## Checklist

 - [ ] ~Requires a [pylibjuju](https://github.com/juju/python-libjuju) change~
 - [ ] ~Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR~
 - [ ] ~Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed~
 - [x] Comments answer the question of why design decisions were made

## QA steps
scale sidecar application to different scale and watch the application status in `juju status`;

```console
$ juju deploy snappass-test

$ juju status --relations --color --storage -m k1:t1
Model  Controller  Cloud/Region        Version  SLA          Timestamp
t1     k1          microk8s/localhost  2.9.9    unsupported  21:19:18+10:00

App            Version  Status  Scale  Charm          Store     Channel  Rev  OS          Address         Message
snappass-test           active      1  snappass-test  charmhub  stable     9  kubernetes  10.152.183.194

Unit              Workload  Agent  Address      Ports  Message
snappass-test/0*  active    idle   10.1.97.236         snappass started

$ juju scale-application snappass-test -m k1:t1 3

$ juju status --relations --color --storage -m k1:t1
Model  Controller  Cloud/Region        Version  SLA          Timestamp
t1     k1          microk8s/localhost  2.9.9    unsupported  21:19:48+10:00

App            Version  Status   Scale  Charm          Store     Channel  Rev  OS          Address         Message
snappass-test           waiting    1/3  snappass-test  charmhub  stable     9  kubernetes  10.152.183.194  installing agent

Unit              Workload  Agent       Address      Ports  Message
snappass-test/0*  active    idle        10.1.97.236         snappass started
snappass-test/1   waiting   allocating  10.1.97.238         installing agent
snappass-test/2   waiting   allocating                      installing agent

$ juju status --relations --color --storage -m k1:t1
Model  Controller  Cloud/Region        Version  SLA          Timestamp
t1     k1          microk8s/localhost  2.9.9    unsupported  21:20:31+10:00

App            Version  Status  Scale  Charm          Store     Channel  Rev  OS          Address         Message
snappass-test           active      3  snappass-test  charmhub  stable     9  kubernetes  10.152.183.194

Unit              Workload  Agent  Address      Ports  Message
snappass-test/0*  active    idle   10.1.97.236         snappass started
snappass-test/1   active    idle   10.1.97.238         snappass started
snappass-test/2   active    idle   10.1.97.237         snappass started

$ juju scale-application snappass-test -m k1:t1 0

$ juju status --relations --color --storage -m k1:t1
Model  Controller  Cloud/Region        Version  SLA          Timestamp
t1     k1          microk8s/localhost  2.9.9    unsupported  21:22:44+10:00

App            Version  Status  Scale  Charm          Store     Channel  Rev  OS          Address         Message
snappass-test           active      0  snappass-test  charmhub  stable     9  kubernetes  10.152.183.194

```

## Documentation changes

No

## Bug reference

https://bugs.launchpad.net/juju/+bug/1934614

### CODE REVIEW

**Review by SimonRichardson — COMMENTED** [review]
This has a lot of schema changes, we should bump OR at the very least understand the damage this could cause by other consumers of the schema.

**Comment by ycliuhw** [comment]
> This has a lot of schema changes, we should bump OR at the very least understand the damage this could cause by other consumers of the schema.

This facade is `only` used for a model manifold worker. Both the client and server side are running together in the same binary.
So I don't think we need to bump version here.

**Comment by ycliuhw** [comment]
[1934614 incorrect k8s sidecar charm app status](https://trello.com/c/OsInWprK/1666-1934614-incorrect-k8s-sidecar-charm-app-status)

File: worker/caasapplicationprovisioner/application.go [hunk 1]
```diff
@@ -257,7 +257,18 @@ func (a *appWorker) loop() error {
 		if done {
```
Comment (hpidcock) [inline worker/caasapplicationprovisioner/application.go:257]: Maybe add a timer to trip this every 10 seconds or so. I don't think we can do a watcher to automatically do this.

**Review by hpidcock — APPROVED** [review]
LGTM works, but a bit slow propagating the status.

**Comment by ycliuhw** [comment]
$$merge$$

**Pull request merged**